kratos/processes/multipoint_constraint_to_element_process.cpp

```diff
@@ -0,0 +1,343 @@
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Máté Kelemen
+//
+
+// Core includes
+#include "processes/multipoint_constraint_to_element_process.h" // MultipointConstraintToElementProcess
+#include "containers/model.h" // Model
+#include "includes/model_part.h" // ModelPart
+#include "includes/master_slave_constraint.h" // MasterSlaveConstraint
+#include "includes/element.h" // Element
+#include "includes/kratos_components.h" // KratosComponents
+#include "includes/key_hash.h" // HashCombine
+#include "utilities/interval_utility.h" // IntervalUtility
+
+// System includes
+#include <unordered_map> // unordered_map
+#include <unordered_set> // unordered_set
+#include <optional> // optional
+
+
+namespace Kratos {
+
+
+struct MultipointConstraintToElementProcess::Impl
+{
+    /// @brief Pair of integers with commutative hash and equality comparisons.
+    template <class TIndex>
+    struct SymmetricIdPair
+    {
+        using value_type = TIndex;
+
+        using first_type = value_type;
+
+        using second_type = value_type;
+
+        struct hash
+        {
+            std::size_t operator()(SymmetricIdPair Instance) const noexcept
+            {
+                value_type min = std::min(Instance.first, Instance.second);
+                const value_type max = std::max(Instance.first, Instance.second);
+                HashCombine(min, max); // <== mutates min
+                return min;
+            }
+        }; // struct hash
+
+        friend bool operator==(SymmetricIdPair Left, SymmetricIdPair Right) noexcept
+        {
+            return (Left.first == Right.first && Left.second == Right.second)
+                || (Left.first == Right.second && Left.second == Right.first);
+        }
+
+        first_type first;
+
+        second_type second;
+    };
+
+
+    /// @brief Pair of symmetric ID pairs.
+    /// @note This pair is NOT symmetric, but the subpairs are.
+    template <class TFirst, class TSecond>
+    struct NestedIdPair
+    {
+        using first_type = SymmetricIdPair<TFirst>;
+
+        using second_type = SymmetricIdPair<TSecond>;
+
+        struct hash
+        {
+            std::size_t operator()(NestedIdPair Instance) const noexcept
+            {
+                auto first_hash = typename first_type::hash()(Instance.first);
+                const auto second_hash = typename second_type::hash()(Instance.second);
+                HashCombine(first_hash, second_hash); // <== mutates first_hash
+                return first_hash;
+            }
+        }; // struct hash
+
+        friend bool operator==(NestedIdPair Left, NestedIdPair Right) noexcept
+        {
+            return Left.first == Right.first && Left.second == Right.second;
+        }
+
+        first_type first;
+
+        second_type second;
+    };
+
+    std::optional<ModelPart*> mpInputModelPart;
```

Comment:
optional of a raw pointer does not make much sense does it?

This is kinda optional of an optional no?
You are not using any `optional` features

Reply:
I'm wrapping `optional`s around pointers to make sure I don't cause segfaults if I forget to check the pointer.

I never assign default-constructed pointers or `nullptr` to the optional, which means if that if the `optional` has a value, it is a valid pointer. On the other hand, if the pointer was never assigned, then nor was the `optional`, in which case an exception is thrown instead of a segfault.